### PR TITLE
Add pkcs11 session pool to PKCS11 LibraryContext.

### PR DESCRIPTION
**Pull request by mykola-kobets-epam**

### CODE REVIEW

File: src/common/pkcs11/pkcs11.cpp [hunk 1]
```diff
@@ -333,7 +333,8 @@ RetWithError<CK_FUNCTION_LIST_PTR> DynamicLibraryContext::Init()
 
 Error LibraryContext::Init()
 {
-    Error err = ErrorEnum::eNone;
+    LockGuard guard(mMutex);
```
Comment (al1img) [inline src/common/pkcs11/pkcs11.cpp:336]: Use name `lock` for this guard and separate if with empty line.

File: src/common/pkcs11/pkcs11.cpp [hunk 2]
```diff
@@ -447,6 +448,7 @@ Error LibraryContext::GetLibInfo(LibInfo& libInfo) const
 
 RetWithError<SharedPtr<SessionContext>> LibraryContext::OpenSession(SlotID slotID, Flags flags)
 {
+    LockGuard     guard(mMutex);
```
Comment (al1img) [inline src/common/pkcs11/pkcs11.cpp:451]: ditto...

File: include/aos/common/pkcs11/pkcs11.hpp [hunk 3]
```diff
@@ -577,22 +577,41 @@ class LibraryContext : public PKCS11LibraryContext {
...
     /**
-     * Opens a session between an application and a token in a particular slot.
+     * Opens a session between an application and a token in a particular slot or returns one from the pool.
      *
      * @param slotID slot identifier.
      * @param flags  indicates the type of session.
      * @return RetWithError<SharedPtr<SessionContext>>.
      */
-    RetWithError<SharedPtr<SessionContext>> OpenSession(SlotID slotID, uint32_t flags);
+    RetWithError<SharedPtr<SessionContext>> OpenSession(SlotID slotID, Flags flags);
```
Comment (al1img) [inline include/aos/common/pkcs11/pkcs11.hpp:586]: Remove extra space in comment `@param flags  indicates the type of session.`

File: tests/common/CMakeLists.txt [hunk 4]
```diff
@@ -34,16 +35,19 @@ set(SOURCES
 )
 
 # ######################################################################################################################
-# Definitions
+# Set PKCS11 test environment
 # ######################################################################################################################
 
-add_definitions(-DAOS_CONFIG_MONITORING_SEND_PERIOD_SEC=2 -DAOS_CONFIG_MONITORING_POLL_PERIOD_SEC=1)
+include(src/pkcs11/pkcs11.cmake)
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply tests/common/CMakeLists.txt:41]: As far as I know the only difference between *.cmake and CMakeLists.txt is that the latter should include any target by itself or transitively via add_subdirectories. In contrary to that current file contains additional functionality that supposed to be uitilized in CMakeLists.txt and doesnt contain any target. Even though it has only a single utilization for pkcs11 unit tests it is made so in order to logically separate target setup from boilerplate test artifacts generation logic. I put it together with artifacts it uses (softhsm.conf) in a separate directory close to a context of use.

File: tests/common/src/pkcs11/pkcs11.cmake [hunk 5]
```diff
@@ -0,0 +1,79 @@
+#
```
[the remark this answers is not in the document]
Comment (mykola-kobets-epam) [reply tests/common/src/pkcs11/pkcs11.cmake:1]: it is used for tests only and I think it should  be placed in the test directory

File: include/aos/common/pkcs11/pkcs11.hpp (hunk 3 above)
Comment (mykola-kobets-epam) [reply include/aos/common/pkcs11/pkcs11.hpp:586]: fixed

File: tests/common/src/pkcs11/pkcs11.cmake [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+# Create PKCS11 test certificates
+# ######################################################################################################################
+
+set(CERTIFICATES_DIR ${CMAKE_CURRENT_BINARY_DIR}/certificates)
+
+file(MAKE_DIRECTORY ${CERTIFICATES_DIR})
+
+message("\nGenerating PKCS11 test certificates...")
+
+message("  Create a Certificate Authority private key...")
```
Comment (al1img) [inline tests/common/src/pkcs11/pkcs11.cmake:39]: Remove leading spaces

File: tests/common/src/pkcs11/pkcs11.cmake [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+execute_process(
+    COMMAND
+        openssl req -new -newkey rsa:2048 -nodes -out ${CERTIFICATES_DIR}/ca.csr -keyout ${CERTIFICATES_DIR}/ca.key
+        -nodes -subj "/C=XX/ST=StateName/L=CityName/O=Epam/OU=CompanySectionName/CN=Aos Core" COMMAND_ERROR_IS_FATAL
+        ANY
+)
+
+message("  Create a CA self-signed certificate:")
```
Comment (al1img) [inline tests/common/src/pkcs11/pkcs11.cmake:47]: ditto

File: tests/common/src/pkcs11/pkcs11.cmake [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+execute_process(
+    COMMAND openssl x509 -trustout -signkey ${CERTIFICATES_DIR}/ca.key -days 365 -req -in ${CERTIFICATES_DIR}/ca.csr
+            -out ${CERTIFICATES_DIR}/ca.pem COMMAND_ERROR_IS_FATAL ANY
+)
+
+message(
+    "  Issue a client certificate by first generating the key, then request, then sign the certificate using private key of the CA..."
```
Comment (al1img) [inline tests/common/src/pkcs11/pkcs11.cmake:54]: Ditto and `Issue a client certificate` message would be enough

File: tests/common/src/pkcs11/pkcs11.cmake [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    COMMAND openssl req -new -key ${CERTIFICATES_DIR}/client.key -out ${CERTIFICATES_DIR}/client.csr -nodes -subj
+            "/C=XX/ST=StateName/L=CityName/O=Epam/OU=CompanySectionName/CN=MKobets" COMMAND_ERROR_IS_FATAL ANY
+)
+
+execute_process(
+    COMMAND openssl x509 -req -days 365 -in ${CERTIFICATES_DIR}/client.csr -CA ${CERTIFICATES_DIR}/ca.pem -CAkey
+            ${CERTIFICATES_DIR}/ca.key -set_serial 01 -out ${CERTIFICATES_DIR}/client.cer COMMAND_ERROR_IS_FATAL ANY
+)
+
+message("  Convert PEM to DER...")
```
Comment (al1img) [inline tests/common/src/pkcs11/pkcs11.cmake:68]: Ditto

File: tests/common/src/pkcs11/pkcs11.cmake (hunk 6 above)
Comment (mykola-kobets-epam) [reply tests/common/src/pkcs11/pkcs11.cmake:39]: fixed

File: tests/common/src/pkcs11/pkcs11.cmake (hunk 7 above)
Comment (mykola-kobets-epam) [reply tests/common/src/pkcs11/pkcs11.cmake:47]: fixed

File: tests/common/src/pkcs11/pkcs11.cmake (hunk 8 above)
Comment (mykola-kobets-epam) [reply tests/common/src/pkcs11/pkcs11.cmake:54]: fixed

File: tests/common/src/pkcs11/pkcs11.cmake (hunk 9 above)
Comment (mykola-kobets-epam) [reply tests/common/src/pkcs11/pkcs11.cmake:68]: fixed

**Review by MykolaSolyanko — APPROVED** [review]

**Pull request merged**